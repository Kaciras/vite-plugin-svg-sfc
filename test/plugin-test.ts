import { cwd } from "process";
import { join, resolve } from "path";
import { writeFileSync } from "fs";
import { expect, it } from "vitest";
import { build, createServer } from "vite";
import { RollupOutput } from "rollup";
import vue from "@vitejs/plugin-vue";
import { createApp } from "vue";
import { renderToString } from "vue/server-renderer";
import { compile, convert, copyFixture, createHMRClient, resolveFixture, useTempDirectory } from "./test-helper.ts";
import svgSfc from "../index.ts";

const input = "image.svg";
const tmpDir = useTempDirectory(cwd());

async function loadBundle<T = any>(code: string) {
	const file = join(tmpDir, "test.js");
	writeFileSync(file, code);
	return (await import("file://" + file)).default as T;
}

it("should throw on non-SVG data", async () => {
	const build = convert("png-data.svg?sfc");
	await expect(build).rejects.toThrow();
});

it("should throw on non-exists file", async () => {
	const build = convert("not exists.svg?sfc");
	await expect(build).rejects.toThrow();
});

it("should keep query in the URL", async () => {
	const bundle = await compile("stroke.svg?foo=1&sfc&bar");
	const absPath = resolveFixture("stroke.svg").replaceAll("\\", "/");
	expect(bundle.output[0].facadeModuleId)
		.toBe(resolve(absPath) + ".vue?foo=1&sfc&bar");
});

it("should not minify on dev mode", async () => {
	expect(await convert("stroke.svg?sfc", { mode: "development" })).toMatchSnapshot();
});

it("should support set minify in options", async () => {
	const code = await convert("stroke.svg?sfc", {
		mode: "production",
		config: { minify: false },
	});
	expect(code).toMatch("<title>Untitled-1</title>");
});

it("should support custom mark", async () => {
	const code = await convert("styles-0.svg?component", {
		config: { mark: "component" },
	});
	expect(code).toMatchSnapshot();
});

it("should work with @vitejs/plugin-vue", async () => {
	const bundle = await build({
		logLevel: "silent",
		build: {
			write: false,
			ssr: resolveFixture("styles-0.svg?sfc"),
		},
		plugins: [vue(), svgSfc()],
	});
	const { code } = (bundle as RollupOutput).output[0];

	const component = await loadBundle(code);
	const app = createApp(component, { width: 4396 });
	expect(await renderToString(app)).toMatchSnapshot();
});

it("should support HMR", async () => {
	const filename = join(tmpDir, "image.svg");
	const mainUrl = `/${input}?sfc`;
	const styleUrl = `/${filename}.vue?vue&type=style&index=0&scoped=true&lang.css`;

	copyFixture("styles-0.svg", filename);

	const server = await createServer({
		logLevel: "silent",
		root: tmpDir,
		build: {
			rollupOptions: {
				input: input + "?sfc",
			},
		},
		plugins: [svgSfc(), vue()],
	});

	async function getStyleCode() {
		await server.transformRequest(mainUrl);
		return (await server.transformRequest(styleUrl))?.code;
	}

	await server.listen();
	const receiveUpdate = createHMRClient(server);

	expect(await getStyleCode()).contains("fill: blue;");

	const waitForHMR = receiveUpdate();
	copyFixture("styles-1.svg", filename);
	const { updates } = await waitForHMR;

	const style = updates.find(e => e.path === styleUrl);
	expect(style?.type).toBe("js-update");
	expect(await getStyleCode()).contains("fill: red;");
});
