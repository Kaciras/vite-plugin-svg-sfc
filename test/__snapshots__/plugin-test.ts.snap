// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should not minify on dev mode 1`] = `"<template><svg id="layer_1" data-name="layer 1" width="1em" height="1em" viewBox="0 0 92.01 92.55" fill="none" stroke="currentColor" stroke-width="2"><title>Untitled-1</title><polygon points="90.02 66.22 58.91 65.25 40.48 90.33 31.79 60.44 2.24 50.66 27.98 33.16 28.15 2.03 52.75 21.1 82.4 11.65 71.86 40.93 90.02 66.22"/></svg></template>"`;

exports[`should support custom mark 1`] = `"<template><svg width="1em" height="1em" fill="currentColor" viewBox="0 0 20 20"><path id="rect" d="M0 0h10v10H0z"/></svg></template><style scoped>#rect{fill:#00f}</style>"`;

exports[`should work with @vitejs/plugin-vue 1`] = `"<svg width="4396" height="1em" fill="currentColor" viewbox="0 0 20 20" data-v-1d0ce38e><path id="rect" d="M0 0h10v10H0z" data-v-1d0ce38e></path></svg>"`;
